File: eco-core/core-nms/v1_17_R1/src/main/java/com/willfp/eco/proxy/v1_17_R1/FastItemStackUtils.java
package com.willfp.eco.proxy.v1_17_R1;

import lombok.experimental.UtilityClass;
import org.bukkit.craftbukkit.libs.org.apache.commons.lang3.Validate;
import org.bukkit.craftbukkit.v1_17_R1.inventory.CraftItemStack;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Field;

@UtilityClass
public class FastItemStackUtils {
    private static final Field field;

    public static net.minecraft.world.item.ItemStack getNMSStack(@NotNull final ItemStack itemStack) {
        if (!(itemStack instanceof CraftItemStack)) {
            return CraftItemStack.asNMSCopy(itemStack);
        } else {
            try {
                return (net.minecraft.world.item.ItemStack) field.get(itemStack);
            } catch (ReflectiveOperationException e) {
                e.printStackTrace();
                return null;
            }
        }
    }

    static {
        Field temp = null;

        try {
            Field handleField = CraftItemStack.class.getDeclaredField("handle");
            handleField.setAccessible(true);
            temp = handleField;
        } catch (ReflectiveOperationException e) {
            e.printStackTrace();
        }

        assert temp != null;
        Validate.notNull(temp, "Error occurred in initialization!");

        field = temp;
    }
}

File: eco-core/core-nms/v1_17_R1/src/main/kotlin/com/willfp/eco/proxy/v1_17_R1/fast/FastItemStackUtils.kt
package com.willfp.eco.proxy.v1_17_R1.fast

import org.bukkit.craftbukkit.v1_17_R1.inventory.CraftItemStack
import org.bukkit.inventory.ItemStack
import java.lang.reflect.Field

class FastItemStackUtils {
    companion object {
        private var field: Field

        init {
            lateinit var temp: Field
            try {
                val handleField = CraftItemStack::class.java.getDeclaredField("handle")
                handleField.isAccessible = true
                temp = handleField
            } catch (e: ReflectiveOperationException) {
                e.printStackTrace()
            }
            field = temp
        }

        fun getNMSStack(itemStack: ItemStack): net.minecraft.world.item.ItemStack {
            return if (itemStack !is CraftItemStack) {
                CraftItemStack.asNMSCopy(itemStack)
            } else {
                field[itemStack] as net.minecraft.world.item.ItemStack
            }
        }
    }
}